### ISSUE TITLE
Pyenv postinstall question

### ISSUE THREAD

**Issue by ryankask** [Homebrew/homebrew#45871]
The pyenv formula currently has the following [postinstall function](https://github.com/Homebrew/homebrew/blob/da9227ea7ad4c0ac15843f8812f4a12efef40eeb/Library/Formula/pyenv.rb) defined:

```ruby
def post_install
  var_lib = HOMEBREW_PREFIX/"var/lib/pyenv"
  %w[plugins versions].each do |dir|
    var_dir = "#{var_lib}/#{dir}"
    mkdir_p var_dir
    ln_sf var_dir, "#{prefix}/#{dir}"
  end

  (var_lib/"plugins").install_symlink "#{prefix}/default-plugins/python-build"
end
```

Is this necessary? What are these directories used for? I don't see anything in the rbenv formula although I don't know how different pyenv is from rbenv.

I only see one file in there and it's a broken symlink:

```shell
$ tree /usr/local/var/lib/pyenv/
/usr/local/var/lib/pyenv/
├── plugins
│   └── python-build
│       └── python-build -> ../../../../Cellar/pyenv/20151105/default-plugins/python-build
└── versions/
```

Please let me know if I can provide any other information.

**Comment by xu-cheng** [comment]
>  it's a broken symlink:

No, it's not.

>  I don't see anything in the rbenv formula

That's because rbenv doesn't bundle with ruby-build, while python-build is included in pyenv.

**Comment by ryankask** [comment]
The symlink is broken on two of the machines I've checked. This happens
after you upgrade pyenv using Homebrew.

I still seem to be able to use `pyenv install` which is why I asked if
`/usr/local/var/lib/pyenv` is needed in the first place.

**Comment by xu-cheng** [comment]
> The symlink is broken on two of the machines I've checked. This happens
after you upgrade pyenv using Homebrew.

Does it remain broken after you run `brew postinstall pyenv`? It doesn't seem to be broken to me.

```
$ tree /usr/local/var/lib/pyenv/
/usr/local/var/lib/pyenv/
├── plugins
│   ├── plugins -> /usr/local/var/lib/pyenv/plugins
│   └── python-build -> ../../../../Cellar/pyenv/20151105/default-plugins/python-build
└── versions
    └── versions -> /usr/local/var/lib/pyenv/versions

5 directories, 0 files

$ ls /usr/local/var/lib/pyenv/plugins/python-build
README.md  bin        install.sh share      test
```

**Comment by ryankask** [comment]
It does remain broken if you don't use `brew cleanup` in between `brew upgrade` and `brew postinstall pyenv` although there is another issues I noticed below.

I'm a casual Homebrew user and I've never had to use `postinstall` before. Also, the other formulas I use don't appear to require such a sequence of commands to avoid cruft being left around.

Anyway, I'm still trying to understand why it's needed in the first place if `pyenv install` appears to be working okay despite the symlink being broken.

***

Note that after running `brew postinstall pyenv` a circular symlink to `/usr/local/var/lib/pyenv/plugins` is added:

```
$ brew postinstall pyenv
$ tree /usr/local/var/lib/pyenv
/usr/local/var/lib/pyenv
├── plugins
│   ├── plugins -> /usr/local/var/lib/pyenv/plugins
│   └── python-build -> ../../../../Cellar/pyenv/20151103/default-plugins/python-build
└── versions
    └── versions -> /usr/local/var/lib/pyenv/versions
```

Please let me know if I'm doing anything wrong.

**Comment by xu-cheng** [comment]
> I've never had to use postinstall before

Normally, Homebrew will run it for you during `brew install`.

> Anyway, I'm still trying to understand why it's needed in the first place if pyenv install appears to be working okay despite the symlink being broken.

You probably need to ask this to upstream. cc @yyuu

**Comment by ryankask** [comment]
> Normally, Homebrew will run it for you during brew install.

Ah, I meant I never had to run it after `brew upgrade`.

Anyway, thanks for your help.

**Comment by yyuu** [comment]
@ryankask The only significant difference between the formulas of `rbenv` and `pyenv` is the existence of `python-build`. The symlink of `default-plugins` is for that purpose.

I have never realized the trouble of upgrading `pyenv` formula. Though there might be something in it. Feel free to fix it if you found out the cause.

**Comment by ryankask** [comment]
@yyuu Thanks for your feedback.

When I installed pyenv with Homebrew in the caveats it says:

```
To use Homebrew's directories rather than ~/.pyenv add to your profile:
  export PYENV_ROOT=/usr/local/var/pyenv
```

The issue discussed above refer to path `/usr/local/var/lib/pyenv`. Note the `lib` directory in `var` -- is that a mistake? I deleted that directory and it my `pyenv` still seems to work fine.

Now my `PYENV_ROOT` is actually `$HOME/.pyenv` so I probably don't need those directories but ideally directories/files that aren't needed wouldn't be created.

It seems like Homebrew is creating files are only needed if the user sets a specific environment variable and when it does create them, they might be wrong?
